crucible-llvm/src/Lang/Crucible/LLVM/MemModel/Strings.hs

```diff
@@ -3,27 +3,261 @@
+concretelyNullTerminatedString =
+  ByteChecker $ \bak acc bytePtr -> do
+    byte <- ptrToBv8 bak bytePtr
+    if isConcreteNullTerminator byte
+    then pure (Break (acc []))
+    else  pure (Continue (acc . (byte:)))
+  where
+    isConcreteNullTerminator symByte =
+      case BV.asUnsigned <$> WI.asBV symByte of
+        Just 0 -> True
+        _ -> False
+
+-- | Helper for loading symbolic strings with a null terminator
+nullTerminatedString ::
+  GHC.HasCallStack =>
+  LCB.IsSymBackend sym bak =>
+  sym ~ WEB.ExprBuilder scope st fs =>
+  bak ~ LCBO.OnlineBackend solver scope st fs =>
+  WPO.OnlineSolver solver =>
+  ByteChecker IO sym bak ([WI.SymBV sym 8] -> [WI.SymBV sym 8]) [WI.SymBV sym 8]
+nullTerminatedString =
+  ByteChecker $ \bak acc bytePtr -> do
+    byte <- ptrToBv8 bak bytePtr
+    let sym = LCB.backendGetSym bak
+    isNullTerm <- isNullTerminator bak sym byte
+    if isNullTerm
+    then pure (Break (acc []))
+    else  pure (Continue (acc . (byte:)))
+  where
+    isNullTerminator bak sym symByte =
+      case BV.asUnsigned <$> WI.asBV symByte of
+        Just 0 -> pure True
+        Just _ -> pure False
+        _ ->
+          LCBO.withSolverProcess bak (pure False) $ \proc -> do
+            z <- WI.bvZero sym (WI.knownNat @8)
+            p <- WI.notPred sym =<< WI.bvEq sym z symByte
+            WPO.checkSatisfiable proc "nullTerminatedString" p <&>
+              \case
+                WS.Unsat () -> True
+                WS.Sat () -> False
+                WS.Unknown -> False
+
+withMaxChars ::
+  GHC.HasCallStack =>
+  LCB.IsSymBackend sym bak =>
+  Functor m =>
+  Int ->
+  (a -> m b) ->
+  ByteChecker m sym bak a b ->
+  ByteChecker m sym bak (a, Int) b
+withMaxChars limit done checker =
+  ByteChecker $ \bak (acc, i) bytePtr ->
+    if i > limit
+    then Break <$> done acc
+    else first (, i + 1) <$> runByteChecker checker bak acc bytePtr
+
+-- | Load a null-terminated string (with a concrete null terminator) from memory.
+--
+-- The string must contain a concrete null terminator. If a maximum number of
+-- characters is provided, no more than that number of characters will be read.
+-- In either case, 'loadConcretelyNullTerminatedString' will stop reading if it
+-- encounters a (concretely) null terminator.
+--
+-- Note that the loaded string may actually be smaller than the returned list if
+-- any of the symbolic bytes may be 0.
+loadConcretelyNullTerminatedString ::
+  ( LCB.IsSymBackend sym bak
+  , Mem.HasPtrWidth wptr
+  , Mem.HasLLVMAnn sym
+  , ?memOpts :: Mem.MemOptions
+  , GHC.HasCallStack
+  ) =>
+  bak ->
+  Mem.MemImpl sym ->
+  Mem.LLVMPtr sym wptr ->
+  -- | Maximum number of characters to read
+  Maybe Int ->
+  IO [WI.SymBV sym 8]
+loadConcretelyNullTerminatedString bak mem ptr limit =
+  case limit of
+    Nothing -> loadBytes bak mem id ptr concretelyNullTerminatedString
+    Just l ->
+      let byteChecker = withMaxChars l (pure . ($ [])) concretelyNullTerminatedString in
```

Comment:
Style nit pick: the expression `pure . ($ [])` is a bit too point-free for my liking. I'd prefer eta-expanding it to `\f -> pure (f [])` to make it a bit clearer what is being applied where.